File: Sources/CountriesFlagsObjC/source/Countries/UIImage+CookIslands.m
//
//  UIImage+CookIslands.m
//  CountriesFlags
//

#import "Countries/UIImage+CookIslands.h"

#import "Extensions/UIBezierPath+Star.h"
#import "Extensions/UIBezierPath+AffineTransform.h"

#import "Countries/UIImage+UnitedKingdom.h"
#import "Extensions/CGSize+Extension.h"

@implementation UIImage (CookIslands)

+ (UIImage * _Nullable)flagCookIslandsWithSize:(CGSize)size {
    UIColor* colorBlue = [UIColor colorWithRed:1.0f/255.0f
                                         green:33.0f/255.0f
                                          blue:105.0f/255.0f
                                         alpha:1.0f];
    UIColor* colorWhite = [UIColor whiteColor];
    
    UIGraphicsBeginImageContextWithOptions(size, YES, 0.0f);
    
    [colorBlue setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorWhite setFill];
    for (int index = 0; index <= 14; index++) {
        UIBezierPath* path = [UIBezierPath starAtPoint:CGPointMake(size.width / 4.0f * 3.0f, size.height / 50.0f * 8.0f)
                                                radius:size.height / 100.0f * 7.5f
                                                  type:FlagStarTypeFivePointed];
        [path rotateToAngle:index * 24.0f
                     center:CGPointMake(size.width / 4.0 * 3.0, size.height / 2.0)];
        [path fill];
    }
    
    [[UIImage flagUnitedKingdomWithSize:CGSizeHalfSize(size)] drawAtPoint:CGPointZero];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image;
}

@end

File: Sources/CountriesFlags/Countries/UIImage+CookIslands.swift
//
//  UIImage+CookIslands.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_the_Cook_Islands "wikipedia"
     
     Generates the flag of the Cook Islands.
     For more information see [wikipedia].
     
     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of the Cook Islands.
     */
    internal static func flagCookIslands(with size: CGSize) -> UIImage? {
        let colorBlue = UIColor(red: 1.0 / 255.0,
                                green: 33.0 / 255.0,
                                blue: 105.0 / 255.0,
                                alpha: 1.0)
        let colorWhite = UIColor.white

        UIGraphicsBeginImageContextWithOptions(size, true, 0.0)

        colorBlue.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorWhite.setFill()
        for index in stride(from: 0.0, through: 14.0, by: 1.0) {
            let path = UIBezierPath.star(at: CGPoint(x: size.width / 4.0 * 3.0, y: size.height / 50.0 * 8.0),
                                         radius: size.height / 100.0 * 7.5)
            path.rotate(to: index * 24.0, around: CGPoint(x: size.width / 4.0 * 3.0, y: size.height / 2.0))
            path.fill()
        }

        flagUnitedKingdom(with: size.halfSize)?.draw(at: .zero)

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image
    }
}

#endif
